networkmanager: Add rc-manager option

talyz opened the pull request:
###### Motivation for this change
Add an option to set the rc-manager parameter in NetworkManager.conf,
which controls how NetworkManager handles resolv.conf. This sets the
default rc-manager to "resolvconf", which solves #61490. It
additionally allows the user to change rc-manager without interference
from configuration activations.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

talyz committed:
networkmanager: Add rc-manager option

Add an option to set the rc-manager parameter in NetworkManager.conf,
which controls how NetworkManager handles resolv.conf. This sets the
default rc-manager to "resolvconf", which solves #61490. It
additionally allows the user to change rc-manager without interference
from configuration activations.

worldofpeace commented on nixos/modules/services/networking/networkmanager.nix:
```diff
@@ -283,6 +284,81 @@ in {
+          <para>
+          Options:
+          <variablelist>
+          <varlistentry>
```
This is pretty much documented at https://developer.gnome.org/NetworkManager/stable/NetworkManager.conf.html
"`main` section" "rc-manager". Can't we link there?

talyz replied:
Sure! For consistency, we should do this for dns as well.

talyz committed:
networkmanager: Documentation cleanup.

- Refer to external documentation for dns option
- Clean up macAddress option
- Improve references

Infinisil commented on nixos/modules/config/networking.nix:
```diff
@@ -309,8 +309,11 @@ in
+        # by NetworkManager
+        ${optionalString (!config.environment.etc?"resolv.conf" &&
+                          cfg.networkmanager.enable ->
+                            cfg.networkmanager.rc-manager == "resolvconf") ''
```
I'm very confused by this condition here, can you explain this?

talyz replied:
My bad, I had the precedence messed up there. It's supposed to check that there's no "resolv.conf"-attribute in etc and that, if networkmanager is enabled, rc-manager is set to "resolvconf".

tobiasBora replied:
I'm also curious, i don't know much about network manager, but after reading the doc it seems that network manager is supposed to run itself the resolvconf program, so I'm not sure why we run here resolvconf. I would have written instead just `config.environment.etc?"resolv.conf" && !cfg.networkmanager.enable`. Did I missed something? Otherwise, thanks for this PR, it solves the problem of no internet after each rebuild!

talyz replied:
NetworkManager runs resolvconf only if configured to do so via `rc-manager = "resolvconf"`, which is why the condition is needed - it makes sure that if NetworkManager is enabled, resolvconf is only run here if NetworkManager is configured to use resolvconf.

etu reviewed (approved):
I reproduced the issue by adding something to my system and doing a `nixos-rebuild test` and watched the `/etc/resolv.conf` loose it's nameservers. So to regain my nameservers I reconnected to my wifi.

Then I applied this patch and did another `nixos-rebuild test` and still had the nameservers in there.

etu commented:
@Infinisil Is there something that you think we need to fix in this after the updates to your comments?

kampka commented:
Any chance we can get this backported to 19.03.
The current behavior using NetworkManager is truly annoying.

Infinisil commented:
@kampka If you want something backported, you're always free to open a PR for it (to the release-19.03 branch)

alyssais commented:
fwiw this is a default behaviour change, so a backport would depend on what the release managers (@samueldr and @lheckemann) think.

lheckemann commented:
-1 on backporting. This can be worked around effectively using

```
networking.networkmanager.extraConfig = ''
 [main]
 rc-manager=resolvconf
'';
```

as mentioned by @talyz on the issue that prompted this, or by defining `networking.nameservers` to be non-empty.
